### PR TITLE
🆕(site) EE console quiz

### LINKED ISSUE #361: Console Quiz
This is an in-console controlled multi-part quiz with:
- a trigger function to start
- console based prompts/questions
- an answer/response function
- session storage for answer results
- award/result response w/ automated action to trigger reward

### PR DESCRIPTION
**Pull request by monkishtypist**
*Dev Tools console based quiz*

This is a console-based quiz with reward and theming functionality.

Resolves #361

### CODE REVIEW

**Comment by monkishtypist** [comment]
@Snugug I wanted you to take a look at this and let me know if this is the right approach you would like to see for exposing JS functions to the dev tools console. Tryna be very eleventy with this approach. Let's open a discussion.

**Comment by monkishtypist** [comment]
@Snugug this has been updated per our conversation in Slack. Functionally all that remains is to trigger the modal and phosphor options on reward.

File: site/js/components/ee.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    if (value.toString().toLowerCase() == this.text.answers[prompt].toString().toLowerCase()) {
+      // Correct
+      console.info('You answered: %c%s', 'font-weight: bold; font-style: italic; background: #0d652d; color: white; padding: .25em;', value.toString());
+      console.info(this.text.prompts[prompt].responseCorrect);
+      this._setCurrentPrompt(prompt + 1);
+      if (this.curPrompt < this.text.prompts.length) {
+        console.info('Question #%s: %c%s', this.curPrompt + 1, 'font-weight: bold;', this._promptQuestion(this.curPrompt, true));
+      } else if ((this.curPrompt = this.text.prompts.length)) {
+        this._promptReward();
+      }
```
Comment (Snugug) [inline site/js/components/ee.js:208]: This looks like it's duplicated from `this._promptQuestion`. Can this all just be replaced with a call to that instead?

File: site/js/components/ee.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+  /**
+   * Check the answer
+   * @param {int} prompt
+   * @param {string} value
+   * @return {void}
+   */
+  _checkAnswer(prompt, value) {
+    // console.info('Question #%s: %c%s', prompt + 1, 'font-weight: bold; font-style: italic;', this._promptQuestion(prompt, true));
+    if (value.toString().toLowerCase() == this.text.answers[prompt].toString().toLowerCase()) {
```
Comment (Snugug) [inline site/js/components/ee.js:199]: With moving answers into the prompt itself, you should be able to grab the `this.text.prompts[prompt]` and reuse that here, for the `responseCorrect`, and `responseIncorrect` instead of needing to construct the lookup each time. It'll also let you pass in the already found prompt to `this._promptHint`

File: site/js/components/ee.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  /**
+   * Set a delay timer
+   * @param {int} milliseconds
+   * @return {void}
+   */
+  _delay(milliseconds) {
+    return new Promise((resolve) => {
+      setTimeout(resolve, milliseconds);
+    });
+  }
```
Comment (Snugug) [inline site/js/components/ee.js:226]: I don't see this used anywhere; am I missing it?

File: site/js/components/ee.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    } else {
+      this._checkAnswer(this.curPrompt, value);
+    }
+  }
+
+  /**
+   * Start over
+   * @return {void}
+   */
+  clear() {
```
Comment (Snugug) [inline site/js/components/ee.js:94]:
```suggestion
  restart() {
```

Clear to me reads as wipe a screen, but not start over. I think this is more obviously "restart from the beginning"

File: site/js/components/ee.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+  /**
+   * Sets the current quiz start value.
+   * @param {int} value - the current quiz start integer value
+   * @return {void}
+   */
+  _setQuizStart(value) {
+    localStorage.setItem('chromeos-m100-quiz-start', value);
+    this.quizStart = value;
+  }
```
Comment (Snugug) [inline site/js/components/ee.js:144]:
```suggestion
  _setStarted(value) {
    localStorage.setItem('chromeos-m100-quiz-start', value);
    this._started = value;
  }
```

File: site/js/components/ee.js [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    }
+  }
+
+  /**
+   * Get started
+   * @return {void}
+   */
+  start() {
+    if (!this.quizStart) {
+      this._setQuizStart(true);
```
Comment (Snugug) [inline site/js/components/ee.js:67]:
```suggestion
      this._setStarted(true);
```

File: site/js/components/ee.js [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        responseIncorrect: "You're almost there!",
+      },
+    ];
+    this.text.answers = ['2011', '50', 'Crostini', '4'];
+    this.text.reward = {
+      headline: 'Huzzah!',
+      body: ["You're officially a Chrome OS insider. We've got more exciting things coming your way for our 100th software release. To stay in the know, sign up for our newsletter, and keep exploring the all-new ChromeOS.dev to help you build your next big idea.", "Thank you for all of your support, we couldn't have reached this milestone without you!", 'The Chrome OS team'],
+    };
+
+    if (!this.quizStart) {
```
Comment (Snugug) [inline site/js/components/ee.js:54]:
```suggestion
    if (!this._started) {
```

File: site/js/components/ee.js [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+      console.warn('%c%s', 'font-weight: bold;', 'm100.start()');
+    } else {
+      this._promptQuestion(this.curPrompt);
```
Comment (Snugug) [inline site/js/components/ee.js:57]:
```suggestion
      this._promptQuestion(this._current);
```

File: site/js/components/ee.js [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+  start() {
+    if (!this.quizStart) {
+      this._setQuizStart(true);
+      console.info('%c%s', 'font-weight: bold; background: #174ea6; color: white; padding: .25em;', this.text.intro.headline);
+      for (const iterator of this.text.intro.body) {
+        console.info(iterator);
+      }
+    }
+    console.warn('Hint: to answer a question use: %c%s', 'font-family: monospace; font-weight: bold;', 'm100.answer(value);');
+    this._promptQuestion(this.curPrompt);
```
Comment (Snugug) [inline site/js/components/ee.js:74]:
```suggestion
    this._promptQuestion(this._current);
```

File: site/js/components/ee.js [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+  /**
+   * Answer
+   * @param {string} value - Submit an answer to a prompted question
+   * @return {void}
+   */
+  answer(value) {
+    if (!this.quizStart) {
```
Comment (Snugug) [inline site/js/components/ee.js:83]:
```suggestion
    if (!this._started) {
```

File: site/js/components/ee.js [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+  /**
+   * Clear the current prompt value.
+   * @return {void}
+   */
+  _clearCurrentPrompt() {
+    localStorage.removeItem('chromeos-m100-current-prompt');
+    this.curPrompt = 0;
+  }
```
Comment (Snugug) [inline site/js/components/ee.js:125]: Why not remove this in favor of setting the current prompt to 0?

File: site/js/components/ee.js [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    this._clearCurrentPrompt();
+  }
+
+  /**
+   * Gets the current prompt value from local storage.
+   * @return {int} - current prompt value
+   */
+  _getCurrentPrompt() {
+    const curPrompt = parseInt(localStorage.getItem('chromeos-m100-current-prompt'));
+    return curPrompt;
```
Comment (Snugug) [inline site/js/components/ee.js:105]:
```suggestion
    return parseInt(localStorage.getItem('chromeos-m100-current-prompt'));
```

**Review by Snugug — CHANGES_REQUESTED** [review]
Thanks for your work here Tim! There are a lot of suggestions here, but they're mostly me just trying to update all of the places where a comment I made actually touched code; the actual comment list is not nearly as large. They can basically be summed up as follows:

* Tweaks to naming to improve clarity and reduce complexity
* Removing some methods in favor of reuse of others
* Reducing the need to look up the current prompt in `_checkAnswer` and removing duplication from that method

Thanks!

File: site/js/components/ee.js [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+   * @return {void}
+   */
+  clear() {
+    this._clearQuizStart();
+    this._clearCurrentPrompt();
+  }
+
+  /**
+   * Gets the current prompt value from local storage.
+   * @return {int} - current prompt value
```
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js]: yup, this is a hangover from my PHP days - oh the glory days ;)

File: site/js/components/ee.js [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+  /**
+   * Check the answer
+   * @param {int} prompt
+   * @param {string} value
+   * @return {void}
+   */
+  _checkAnswer(prompt, value) {
+    // console.info('Question #%s: %c%s', prompt + 1, 'font-weight: bold; font-style: italic;', this._promptQuestion(prompt, true));
```
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js:204]: I was holding onto this in case we wanted to re-introduce the question upon submitting an answer. But this can be re-added later if needed. removed.

File: site/js/components/ee.js (hunk 3 above)
Comment (monkishtypist) [reply site/js/components/ee.js]: I was playing with adding time delays to text rendering. However this felt clunky at times, so I removed those delays. I left the function in case we wanted to re-examine this use case.

File: site/js/components/ee.js [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+/**
+ * Bootstrap the M100 Console Quiz
+ */
+export class M100 {
+  constructor() {
+    this.curPrompt = this._getCurrentPrompt() || 0;
+    this.quizStart = this._getQuizStart();
+
+    this.text = {};
```
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js:17]: lolz, I did it that way, then combined it all (shrug), and am now reverting.

File: site/js/components/ee.js (hunk 4 above)
Comment (monkishtypist) [reply site/js/components/ee.js:97]: agreed +1 - this was originally just clearing vars for testing, but I find might be useful for the user to start over.

File: site/js/components/ee.js [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+ * Need a HELP function - to show question, available commands etc.
+ */
+
+/**
+ * Bootstrap the M100 Console Quiz
+ */
+export class M100 {
+  constructor() {
+    this.curPrompt = this._getCurrentPrompt() || 0;
+    this.quizStart = this._getQuizStart();
```
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js:15]: sure. I'm from the school of using naming _as_ documentation (eli5 naming), but quickly updated.

File: site/js/components/ee.js [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+
+  /**
+   * Prompt the hint to the user
+   * @param {int} prompt
+   */
+  _promptHint(prompt) {
+    if (this.text.prompts[prompt].hint.length > 0) {
+      console.info('Hint: %c%s', 'font-style: italic;', this.text.prompts[prompt].hint);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js:184]: good call out. updated.

File: site/js/components/ee.js (hunk 1 above)
Comment (monkishtypist) [reply site/js/components/ee.js:214]: good catch. I had different text strings before, but normalized them.

File: site/js/components/ee.js [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+
+  /**
+   * Clear the current quiz start value.
+   * @return {void}
+   */
+  _clearQuizStart() {
+    localStorage.removeItem('chromeos-m100-quiz-start');
+    this.quizStart = false;
+  }
```
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js]: Since we are using localstorage I wanted to isolate these methods and allow for future extensibility in case we need to clear other things tied to start (like a maybe a step indicator or theme class setting or etc)

File: site/js/components/ee.js (hunk 11 above)
Comment (monkishtypist) [reply site/js/components/ee.js]: Same here I wanted to isolate these methods and allow for future extensibility in case we need to clear other things tied to current step (something we haven't thought of yet?)

File: site/js/components/ee.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (monkishtypist) [reply site/js/components/ee.js]: After new flow review, these are being removed

File: site/js/components/ee.js [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    if (value.toString().toLowerCase() == prompts.answer.toString().toLowerCase()) {
+      // Correct
+      console.info('You answered: %c%s', 'font-weight: bold; font-style: italic; background: #0d652d; color: white; padding: .25em;', value.toString());
+      console.info(prompts.responseCorrect);
+      this._setCurrent(prompt + 1);
+      if (this._current < this._prompts.length) {
+        this._promptQuestion(this._current);
+      } else if ((this._current = this._prompts.length)) {
+        this._promptReward();
+      }
```
Comment (Snugug) [inline site/js/components/ee.js:211]: Doesn't `promptQuestion` have the `promptReward` logic included in it?

File: site/js/components/ee.js [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    } else {
+      this._checkAnswer(this._current, value);
+    }
+  }
+
+  /**
+   * Start over
+   * @return {void}
+   */
+  restart() {
```
Comment (Snugug) [inline site/js/components/ee.js:96]:
```suggestion
  reset() {
```

I think with our conversation this afternoon, maybe switching this to `reset` is better

**Review by Snugug — CHANGES_REQUESTED** [review]
Thanks for the updates here! Just two comment, one based on our conversation this afternoon.

File: site/js/components/ee.js (hunk 19 above)
Comment (monkishtypist) [reply site/js/components/ee.js]: this has been re-refined.

File: site/js/components/ee.js (hunk 20 above)
Comment (monkishtypist) [reply site/js/components/ee.js]: agreed!

File: site/js/worklets/theme.js [hunk 21]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Site Theme setter and getter
+ */
+export class Theme {
```
Comment (Snugug) [inline site/js/worklets/theme.js:4]: `worklets` have a very specific meaning; it's a [light-weight Web Worker used as part of the rendering pipeline](https://developer.mozilla.org/en-US/docs/Web/API/Worklet). This appears to be a component like our other components. Please move and rename as appropriate

File: site/js/main.js [hunk 22]
```diff
@@ -57,6 +58,13 @@ window.addEventListener('DOMContentLoaded', async () => {
     offlineSearch.style.display = 'block';
   }
 
+  if (body) {
+    const { Theme } = await import('./worklets/theme');
+    const theme = new Theme(body);
+    const storedTheme = theme._getStoredTheme();
+    theme.name = storedTheme;
+  }
```
Comment (Snugug) [inline site/js/main.js:66]: A couple of things here:
* Appreciate the re-use of our conditional checking pattern here, but `body` is always going to be available, so we don't need to add a conditional in for it. We also don't need to pass it in; it's a known entity that we can grab directly (aka doesn't require querying for)
* I don't see `storedTheme` or `theme` being used outside of this initialization. If they're not, we can just do `new Theme()`

File: site/js/worklets/theme.js [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   *
+   * @param {DOMElement} elem
+   * @param {string} name
+   * @return {void}
+   */
+  constructor(elem, name) {
+    this._elem = elem;
```
Comment (Snugug) [inline site/js/worklets/theme.js:12]: Because of the way `name` works, I don't see this being usable by other elements. We should just use `document.body` directly here

File: site/js/components/ee.js [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  /**
+   * Toggle ChromeOS theme
+   * @param {string} theme
+   * @return {void}
+   */
+  _setTheme(theme = null) {
+    localStorage.setItem('chromeos-theme', theme);
+  }
```
Comment (Snugug) [inline site/js/components/ee.js:226]: I feel like this is a good place to re-use the `theme` class and methods that were made in elsewhere in this work instead of re-defining the same functionality here.

File: site/js/components/ee.js [hunk 25]
```diff
@@ -0,0 +1,250 @@
+/**
+ * Bootstrap the M100 Console Quiz
+ */
+export class M100 {
+  /**
+   *
+   * @return {void}
+   */
+  constructor() {
+    this._body = document.querySelector('#body');
```
Comment (Snugug) [inline site/js/components/ee.js:10]: This goes along with my other theme related comments

File: site/js/components/ee.js [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+      console.info('Hint: %c%s', 'font-style: italic;', hint);
+    }
+  }
+
+  /**
+   * Prompt reward
+   * @param {boolean} giveYouUp
+   * @return {void}
+   */
+  _promptReward(giveYouUp = false) {
```
Comment (Snugug) [inline site/js/components/ee.js:178]: I'm not seeing an instance where this is called without it being `true`. Do we need to have the param (and follow-up if statement) here?

File: site/js/components/ee.js [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+  _checkAnswer(current, value) {
+    const prompts = this._prompts[current];
+    if (value.toString().toLowerCase() == prompts.answer.toString().toLowerCase()) {
+      // Correct
+      console.info('You answered: %c%s', 'font-weight: bold; font-style: italic; background: #0d652d; color: white; padding: .25em;', value.toString());
+      console.info(prompts.responseCorrect);
+      this._setCurrent(current + 1);
+      if (this._current < this._prompts.length) {
+        this._promptQuestion(this._current);
+      } else if ((this._current = this._prompts.length)) {
```
Comment (Snugug) [inline site/js/components/ee.js:208]: This is not a comparison; is that intended?

File: site/js/components/ee.js [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+    for (const iterator of this._reward.body) {
+      console.info(iterator);
+    }
+    this._setTheme('phosphor');
+    if (giveYouUp) {
+      const delay = 3;
+      this._countdown(delay);
+      setTimeout(() => {
+        this._giveYouUp();
+      }, delay * 1000);
```
Comment (Snugug) [inline site/js/components/ee.js:189]: I think passing in a callback here, instead of relying on `setTimeout` scheduling, would be better here.

File: site/js/worklets/theme.js [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   *
+   * @param {string} newName
+   * @return {void}
+   */
+  set name(newName) {
+    newName = newName.trim();
+    this._name = newName;
+    this._elem.setAttribute('data-theme', newName);
+    localStorage.setItem('chromeos-theme', newName);
```
Comment (Snugug) [inline site/js/worklets/theme.js:32]:
```suggestion
   * @param {string} value
   * @return {void}
   */
  set name(value) {
    this._name = value.trim();
    this._elem.setAttribute('data-theme', this._name);
    localStorage.setItem('chromeos-theme', this._name);
```

Comment (Snugug) [reply site/js/worklets/theme.js:32]: Simpler naming

**Review by Snugug — CHANGES_REQUESTED** [review]
Thanks for the update here! A couple more comments, here's the summary:

* Prefer `document.body` as it's a direct, known, standardized way of getting the body of a document, instead of querying for it
* Theme isn't a worklet, so it should be moved and renamed/commented
* `ee.js` seems to be duplicating functionality in Theme. It'd be better if it re-used the functionality available in Theme.

Thanks!

File: site/js/worklets/theme.js (hunk 23 above)
Comment (monkishtypist) [reply site/js/worklets/theme.js:12]: I see it. updated.

File: site/js/worklets/theme.js (hunk 21 above)
Comment (monkishtypist) [reply site/js/worklets/theme.js:4]: moved to `components`

File: site/js/components/ee.js (hunk 24 above)
Comment (monkishtypist) [reply site/js/components/ee.js:226]: updated.

File: site/js/components/ee.js (hunk 27 above)
Comment (monkishtypist) [reply site/js/components/ee.js:208]: typo!! updated.

File: site/js/components/ee.js (hunk 28 above)
Comment (monkishtypist) [reply site/js/components/ee.js:189]: yupdated!

File: site/js/components/ee.js (hunk 26 above)
Comment (monkishtypist) [reply site/js/components/ee.js:178]: lol. legacy testing. I kept rick-rolling myself...

File: site/js/worklets/theme.js (hunk 29 above)
Comment (monkishtypist) [reply site/js/worklets/theme.js:32]: updated

File: site/js/components/ee.js [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+ * Bootstrap the M100 Console Quiz
+ */
+export class M100 {
+  /**
+   *
+   * @return {void}
+   */
+  constructor() {
+    this._theme = new Theme();
+    this._body = document.body;
```
Comment (Snugug) [inline site/js/components/ee.js:16]:
```suggestion
```

This doesn't appear to be used

**Review by Snugug — DISMISSED** [review]

File: site/js/components/ee.js (hunk 30 above)
Comment (monkishtypist) [reply site/js/components/ee.js]: oops.

**Review by Snugug — APPROVED** [review]
